src/backend/columnar/columnar_tableam.c

```diff
@@ -253,13 +251,38 @@ CreateColumnarScanMemoryContext(void)
  */
 static ColumnarReadState *
 init_columnar_read_state(Relation relation, TupleDesc tupdesc, Bitmapset *attr_needed,
-						 List *scanQual, MemoryContext scanContext)
+						 List *scanQual, MemoryContext scanContext, Snapshot snapshot)
 {
 	MemoryContext oldContext = MemoryContextSwitchTo(scanContext);
 
+	Oid relfilenode = relation->rd_node.relNode;
+	FlushWriteStateForRelfilenode(relfilenode, GetCurrentSubTransactionId());
+
+	bool snapshotCreatedByUs = false;
+	if (snapshot != InvalidSnapshot && IsMVCCSnapshot(snapshot))
+	{
+		/*
+		 * If we flushed any pending writes, then we should guarantee that
+		 * those writes are visible to us too. For this reason, if given
+		 * snapshot is an MVCC snapshot, then we set its curcid to current
+		 * command id.
+		 *
+		 * For simplicity, we do that even if we didn't flush any writes
+		 * since we don't see any problem with pushing a new snapshot.
+		 */
+		PushCopiedSnapshot(snapshot);
+
+		/* now our snapshot is the active one */
+		UpdateActiveSnapshotCommandId();
+		snapshot = GetActiveSnapshot();
+
+		snapshotCreatedByUs = true;
```

Comment:
It's a little hard for me to follow what's happening here. If we aren't using GetActiveSnapshot() anywhere, why not just pop it unconditionally after updating the command ID? And I don't think EndRead is guaranteed to be called, so that means we can end up leaving a snapshot on the stack for no reason.

And if using the active stack is useful, why not always use it, so we don't have to keep track of whether to pop it or not? That would make one code path rather than two, which is more likely to be exercised thoroughly.

Reply:
> hy not just pop it unconditionally after updating the command ID? 

Agree, this makes a lot of sense, thank you.

Reply:
Then I guess pushing our snapshot to stack and not keeping it in readState is the cleanest way ?

But then, if EndRead is not guaranteed to be called, then we should call `PopActiveSnapshot` immediately but keep the snapshot in the readState.

However, even if we do that, I think we should make sure to Register / Unregister snapshot in BeginRead / EndRead. That again requires keeping snapshot in readState, but then what happens if EndRead doesn't get called ?

So I'm not sure how to proceed from here.

Reply:
It looks like ColumnarEndRead will be callled even if the executor node doesn't run to completion (e.g. limit or exists query). I thought I remembered a case where it might not be called, but I'm not seeing it now.

So the Push/Update/Pop/Register approach seems best. Let me know if you run into a problem case.